cores/nRF5/HardwarePWM.cpp

```diff
@@ -350,3 +270,58 @@ uint8_t HardwarePWM::freeChannelCount(void) const
   return MAX_CHANNELS - usedChannelCount();
 }
 
+// returns true ONLY when (1) no PWM channel has a pin, and (2) the owner token is nullptr
+bool HardwarePWM::takeOwnership(uintptr_t token)
+{
+  if (token == 0) {
+    LOG_LV1("HwPWM", "zero / nullptr is not a valid ownership token (attempted use in takeOwnership)");
+    return false;
+  }
+
+  if (token == this->_owner_token) {
+    LOG_LV1("HwPWM", "failing to acquire ownership because already owned by requesting token (cannot take ownership twice)");
+    return false;
+  }
+
+  if ( this->_owner_token != 0 ) return false;
+  if ( this->usedChannelCount() != 0 ) return false;
+  if ( this->enabled() ) return false;
+
+  // This function must not be called within ISR
+  taskENTER_CRITICAL();
+  _owner_token = token;
```

Comment:
This is not correct ... you have introduced a bug.
The old code:

```C++
uintptr_t newValue = 0U;	
return this->_owner_token.compare_exchange_strong(newValue, token);
```

In contrast, the code inside the critical section here is modifying the value unconditionally ... a clear race condition.

Reply:
I don't see anything except SoftDevice could do a race condition here. SD is clearly won't run be able to do a race condition here.

Reply:
#### _Yes, there is a race condition._

Your code is not equivalent.  Read the definition of `compare_exchange_strong()` to understand.

#### Example race condition

1. Start with `_owner_token == nullptr`
2. Task 1: (low priority task) runs to line 290 quoted above
3. ISR: interrupt fires, pre-empting Task 1
4. ISR: calls the function with owner token 0xFEED, which **_succeeds_**
    * `_owner_token == 0xFEED`
    * Note: ISR could also have caused a higher-priority task to execute, which called the function...
5. Task 1: Resumes operation at line 290
6. Task 1: sets _owner_token to 0xBAADF00D (overwrites the existing owner 0xFEED)
    * `_owner_token == 0xBAADF00D`
7. Task 1: returns an indication of **_success_**

Two tasks believe they successfully obtained ownership, and then corrupt device state.

Reply:
that is mostly true but only without critical section. https://www.freertos.org/taskENTER_CRITICAL_taskEXIT_CRITICAL.html 
there is no ISR that can interrupt the task once the API is invoked.

Reply:
Ah I see the racing scenario that you mentioned, we can move the code a bit to fix the racing.